cdap-security/src/test/java/co/cask/cdap/security/authorization/InMemoryAuthorizer.java

```diff
@@ -33,25 +33,35 @@
 import com.google.common.collect.Sets;
 import com.google.common.collect.Table;
 
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
-import javax.annotation.concurrent.NotThreadSafe;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 /**
  * In-memory implementation of {@link Authorizer}.
  */
-@NotThreadSafe
 public class InMemoryAuthorizer extends AbstractAuthorizer {
 
   private final Table<EntityId, Principal, Set<Action>> table = HashBasedTable.create();
-  private final Map<Role, Set<Principal>> roleToPrincipals = new HashMap<>();
+  private final Map<Role, Set<Principal>> roleToPrincipals = new ConcurrentHashMap<>();
```

Comment:
However, judging from the usage of the `roleToPrincipals`, there are race conditions. E.g. in `createRole`:

```java
     if (roleToPrincipals.containsKey(role)) {
       throw new RoleAlreadyExistsException(role);
     }
     // NOTE: A concurrent put might happen, hence it should still result as RoleAlreadyExistsException.
     // Better just do a putIfAbsent call, and throw exception if failed to put.
     roleToPrincipals.put(role, new HashSet<Principal>());
```